go/chat/maps/livelocation.go

```diff
@@ -0,0 +1,403 @@
+		l.saveLocked(ctx)
+	}()
+
+	if !t.getCurrentPosition {
+		// if this is a live location request, just put whatever the last coord is on the screen, makes it
+		// feel more live
+		if !l.lastCoord.IsZero() {
+			t.updateCh <- l.lastCoord
+		}
+	}
+	firstUpdate := true
+	shouldUpdate := false
+	nextUpdate := l.clock.Now().Add(l.updateInterval)
+	for {
+		select {
+		case coord := <-t.updateCh:
+			if !t.getCurrentPosition {
+				t.Drain(coord)
+				shouldUpdate = true
+				if firstUpdate {
+					l.Debug(ctx, "tracker: updating due to live location first update")
+					l.updateMapUnfurl(ctx, t)
+				}
+				firstUpdate = false
+			} else {
+				t.SetCoords([]chat1.Coordinate{coord})
+			}
+			l.Lock()
+			l.saveLocked(ctx)
+			l.Unlock()
+		case <-l.clock.AfterTime(nextUpdate):
+			// we update the map unfurl on a timer so we don't spam delete and recreate it
+			if shouldUpdate {
+				// drain anything in the buffer if we are being updated and posting at the same time
+				t.Drain(chat1.Coordinate{})
+				l.Debug(ctx, "tracker: updating due to next update")
+				l.updateMapUnfurl(ctx, t)
+				shouldUpdate = false
+			}
+			nextUpdate = l.clock.Now().Add(l.updateInterval)
+		case <-l.clock.AfterTime(t.endTime):
+			l.Debug(ctx, "tracker: live location complete: watchID: %d", watchID)
+			if t.getCurrentPosition || shouldUpdate {
+				t.Drain(chat1.Coordinate{})
+				l.Debug(ctx, "tracker: updating due to expiration")
+				l.updateMapUnfurl(ctx, t)
+			}
+			return nil
+		case <-t.stopCh:
+			return nil
+		}
+	}
+}
+
+func (l *LiveLocationTracker) GetCurrentPosition(ctx context.Context, convID chat1.ConversationID,
+	msgID chat1.MessageID) {
+	defer l.Trace(ctx, func() error { return nil }, "GetCurrentPosition")()
+	l.Lock()
+	defer l.Unlock()
+	// start up a live location tracker for a small amount of time to make sure we get a good
+	// coordinate
+	t := newLocationTrack(convID, msgID, l.clock.Now().Add(4*time.Second), true, l.maxCoords)
+	l.trackers[t.Key()] = t
+	l.saveLocked(ctx)
+	l.eg.Go(func() error { return l.tracker(t) })
+}
+
+func (l *LiveLocationTracker) StartTracking(ctx context.Context, convID chat1.ConversationID,
+	msgID chat1.MessageID, endTime time.Time) {
+	defer l.Trace(ctx, func() error { return nil }, "StartTracking")()
+	l.Lock()
+	defer l.Unlock()
+	t := newLocationTrack(convID, msgID, endTime, false, l.maxCoords)
+	l.trackers[t.Key()] = t
+	l.saveLocked(ctx)
+	l.eg.Go(func() error { return l.tracker(t) })
+}
+
+func (l *LiveLocationTracker) LocationUpdate(ctx context.Context, coord chat1.Coordinate) {
+	defer l.Trace(ctx, func() error { return nil }, "LocationUpdate")()
+	l.Lock()
+	defer l.Unlock()
+	if l.G().IsMobileAppType() {
+		// if the app is woken up as the result of a location update, and we think we are currently
+		// backgrounded, then go ahead and mark us as background active so that we can get
+		// location updates out
+		l.G().MobileAppState.UpdateWithCheck(keybase1.MobileAppState_BACKGROUNDACTIVE,
```

Comment:
can there be a new `MobileAppState_BACKGROUNDACTIVELOCATION` state? i worry about other background syncing coming online while doing the location updates and draining battery/data

Reply:
Can do that in a subsequent PR if we think it is useful. The main reason we need this is for various components in chat that require the app to be running (not in the background) to do stuff. I'd have to track all that down and pick out only the ones we need for location. I'm not sure that is worth it right now. The biggest battery drain of this feature is keeping the GPS powered on anyway.

Reply:
ok cool we can keep on eye on it